Interpolate weecology esss

juniperlsimonis opened the pull request:
this PR adds a new model script for "weecology-esss", the provisional more specific name of "naive01", which includes interpolation for missing data. 
this model is not directly comparable to the four existing models as the data set is augmented to account for the missing values

juniperlsimonis committed:
created weecology-esss

this model file will include the interpolated esss model, also with an updated name

juniperlsimonis committed:
header

juniperlsimonis committed:
main function

juniperlsimonis committed:
application code

juniperlsimonis committed:
interpolating individual species then adding

handling jensens inequality in the interpolation

juniperlsimonis committed:
small edits

align argument inputs
suppressing warnings associated with changing factors  to characters to combine them

juniperlsimonis committed:
internal renaming of some objects

ethanwhite commented on models/weecology-esss.R:
```diff
@@ -0,0 +1,155 @@
+
+  # Combine output
+  #  warnings are suppressed here because they're just associated with 
+  #  coercing un-matching sets of factors to character vectors to bind
```
My temptation here is to avoid suppressing the warnings by changing the factors into character vectors somewhere. This removes the need to the comment and the (sort of scary) `suppressWarnings` calls.

DAPPERstats replied:
good call. i definitely don't like to use suppressWarnings

ethanwhite commented on models/weecology-esss.R:
```diff
@@ -0,0 +1,155 @@
+            file.path('tmp', 
+                paste("Weecology-ESSS", filename_suffix, "_model_aic.csv", 
+                      sep = "")),
+            row.names = FALSE)
```
This sort of multi-level indenting for line length can make code difficult to read. In this case I think that pulling the filename creation out might make it more readable, e.g.,

```
fcast_path <- file.path('tmp', paste("Weecology-ESSS", filename_suffix, ".csv", sep = ""))
write.csv(forecasts, fcast_path, row.names = FALSE)
```

Once the "Weecology-" is pulled this fits in 80 chars without the challenging read.

ethanwhite commented on models/weecology-esss.R:
```diff
@@ -0,0 +1,155 @@
+    forecast_years <- model_metadata$forecast_years
+    forecast_newmoons <- model_metadata$forecast_newmoons
+
+  # Forecast All plots
```
The content of this content should be captured in the file name and the comment removed. If the current `allresults` doesn't quite do that for you then `all_forecasts` or `forecasts_all` or something.

You can also make things more explicit by use verbs for function calls. E.g.,

```
all_forecasts <- forecast_ess(...
```

or

```
all_forecasts <- get_ess_forecasts(...
```

ethanwhite commented on models/weecology-esss.R:
```diff
@@ -0,0 +1,155 @@
+
+    # fit the ets model and forecast with it
+
+      etsm <- ets(interpolated_total)
```
I'm be a bit more expansive with the variable names here. Since `etsm` doesn't mean anything to my brain I have to read the code t the right, understand that this is fitting a model named `ets` and then infer that the `m` means model. In contrast `ets_model` is more easily understandable. I'd then change the variable name on the following line to `ets_forecast` or `ets_fcast`.

ethanwhite commented on models/weecology-esss.R:
```diff
@@ -0,0 +1,155 @@
+                          level = CI_level, 
+                          allow.multiplicative.trend = TRUE)
+
+    # prep the forecast data tabe
```
Drop the comment and capture the information in the object name. I'd probably trying something like `output`, `fcast_output`, or `forecast_output`. This are varying levels of detail but all quickly readable.

ethanwhite reviewed (changes requested):
This looks really good! The code is in great functional shape so I'm going to make a bunch of suggestions aimed at improving the overall style. This all applies to (and interacts with) #197, so apply these ideas to both PRs and we can go from there. If it's easier to just send one combined PR at that point that's fine too.

1. Let's pull the interpolation code out into a function, put it in a file called `tools/model_tools.R` where shared tools for models can be stored, and source it into all of the models that need it. This will keep code duplication to minimum, let us document the details of the interpolation, and still make the choice to interpolate model specific. For flexibility I'd probably have that function interpolate for all species and then do the species level filtering in the model scripts.

2. If we're envisioning this as a modification/replacement of the previous model then I think we should go ahead and remove the other model as part of this PR. @gmyenni - does that sound right to you?

3. I think adding "weecology" to names is getting ahead of ourselves. It would be great if so many folks are contributing (similar) models that we have to deal with naming issues, but in reality that's unlikely and I think it's better to deal with that if it becomes an issue. This naming will be confusing and we should only add that level of confusion if necessary.

4. The (undocumented; sorry) lab code style is `_` not `.` for separating words names. In general we follow http://adv-r.had.co.nz/Style.html

gmyenni commented:
Yeah, I think we should remove the two 'naive' models and replace with these.

Regarding model naming:

If this is considered an improvement of 'naive01', then we want the model name to stay the same, so all the predictions get evaluated together, right? Right now the model name is `Forecast`, which is terrible and should definitely be changed to `ESSS`. But we'll have to add some patchy code to collect old `Forecast` and new `ESSS` predictions together, or update the model name in all the old predictions.

The same issue over in #197 is easier. The model name of 'naive02' is AutoArima. So if you remove the 'naive02' model script, and use `model = AutoArima` instead of arima, that will seamlessly replace it.

If these are replacements that should be evaluated separately, then the model names should not stay the same. In which case, there's no issue with naming for this model; `Forecast` and `ESSS` will get evaluated separately. But the new arima should probably get a more specific name, to distinguish it from the existing `AutoArima`.

DAPPERstats commented:
thanks for the feedback!
totally on board with the editing changes to things. those all make good sense. 

totally on board with pulling out the interpolation function. that makes 100% sense.

the issue with replacement of the models is a bit confusing to me because of the fact that the new models are using a different set of data (interpolated) than the existing models (not interpolated). so they should not be directly compared in any real way. they are not being fit using the same data and technically are not forecasting the same data as a result.

ethanwhite commented:
> the issue with replacement of the models is a bit confusing to me because of the fact that the new models are using a different set of data (interpolated) than the existing models (not interpolated). so they should not be directly compared in any real way. they are not being fit using the same data and technically are not forecasting the same data as a result.

Here's my and @skmorgane's take on this. Certainly in some cases these models aren't comparable (i.e., the AICs), but we think they are directly comparable when it comes to the forecasts themselves. As long as we are explicit that something is a forecast for value *y* at point in time *t* it doesn't matter how we got it. We could use professional judgement, a model that says it's a Poisson sample of the last observed time point with no fitting, or we could fit a model to data - either interpolated or not.

Assuming we all agree on this general point then the comparability issue is for the AIC values used for ensembling and potentially for how we're assessing model performance in forecasting/hindcasting. For the AIC based model weighting we'd ideally like the training data to be the same across models (though there are better ways to do this weighting that would remove this issue down the line). We can't compare the AIC values through time since data points are changing so the added inability to compare them due to the change in interpolation doesn't stop us from doing anything. Forecasts and hindcasts should only be evaluated on their predictions for actually sampled data points, not interpolated ones.

So, our take is that these changes should be considered improvements to the current models that are known to have occurred at a specific point in time and can be assessed when we look back at how performance has changed (based on evaluation against future data, not the AIC values). We should rename everything to the best possible names. We should then update all of the old forecast output files to follow the same naming convention (separate PR, probably by @gmyenni once the new models). This will give us a better version of these models that can easily be compared to the previous versions without having to remember to bootstrap the old naming conventions.

This should also make updating the `tscount` models pretty easy since it should basically be adding the new interpolation function to swap out the abundance data for the interpolated data.

> sorry, I suck at naming things ¯_(ツ)_/¯)

https://www.martinfowler.com/bliki/TwoHardThings.html

juniperlsimonis committed:
removal of nave01 script

new model is replacing this one

juniperlsimonis committed:
renamed main model function file

removed the weecology- lead in

juniperlsimonis committed:
renaming the new model within the script

in function name and output values

juniperlsimonis committed:
moved interpolation code into its own function

now in the (new) model_tools script
note that there's a lot hard-coded and assumed in the interpolation function right now that can easily be relaxed to work on other versions of the data (for example, working with plot-level data or only a subset of species or a different taxonomic group)

juniperlsimonis committed:
add source line to main forecast script

juniperlsimonis committed:
formatting and naming edits to the main function

juniperlsimonis committed:
formatting and name editing

DAPPERstats commented:
the code here's been updated with edits to address your comments and should be ready to get merged 🤞 once the build checks are clear

juniperlsimonis committed:
edit to model output format

i tried previously to make everything a character but downstream stuff needs particular formatting for numerics, so i am now just turning the factor columns to characters and then combining

juniperlsimonis committed:
slight name edit

DAPPERstats commented:
ok, checks have passed!
ready to be evaluated for merging.

juniperlsimonis committed:
include print to show running

to align with the GARCH models

ethanwhite commented on models/esss.R:
```diff
@@ -0,0 +1,111 @@
+                               fit_end_newmoon = fit_end_newmoon,
+                               initial_newmoon = initial_newmoon)
+    for(i in 5:8)
+      output_fcast[ , i] <- as.character(output_fcast[ , i])
```
Could this be replaced by a `stringsAsFactors = FALSE` in the data.frame definition?

DAPPERstats replied:
yup! good call

ethanwhite commented on models/esss.R:
```diff
@@ -0,0 +1,111 @@
+                             fit_end_newmoon = fit_end_newmoon,
+                             initial_newmoon = initial_newmoon)
+    for(i in 2:5)
+      output_aic[ , i] <- as.character(output_aic[ , i])
```
Same question about `stringsAsFactors = FALSE`

ethanwhite commented on models/esss.R:
```diff
@@ -0,0 +1,111 @@
+    for(i in 5:8)
+      output_fcast[ , i] <- as.character(output_fcast[ , i])
+    for(i in 9:11)
+      output_fcast[ , i] <- as.numeric(output_fcast[ , i])
```
Does this conversion from time-series to numeric influence the output/warnings? If not then I'd drop this. If so then I'd probably add the `as.numeric`'s into the data frame definition instead. The idea being that

```
estimate = as.numeric(ets_forecast$mean)
```

reads pretty clearly as "Define estimate as ets_forecast$mean while addressing a type issue", where as this loop requires more thought - "OK, they just defined a data frame, and they are looping over it to do what... OK, changing type... So, which columns are 9:11..."

DAPPERstats replied:
yeah, it's to change the predictions from class ts to numeric
I like your suggestion here, although to avoid long lines, I'd suggest we pull the definition out to before the data frame definition

`estimate <- as.numeric(ets_forecast$mean)`
`output_fcast <- data.frame(..., estimate = estimate, ...)`

let me know if you'd see any issues with that

ethanwhite replied:
👍

ethanwhite commented on models/esss.R:
```diff
@@ -0,0 +1,111 @@
+  aics <- rbind(forecasts_all[[2]], forecasts_controls[[2]])
+
+  fcast_path <- paste("ESSS", filename_suffix, ".csv", sep = "")
+  fcast_path <- file.path('tmp', fcast_path)
```
I'd probably put these all on one line (shortening `filename_suffix` if needed for length):

```
fcast_path <- file.path('tmp', paste("ESSS", filename_suffix, ".csv", sep = ""))
```

If we want to keep them on separate lines the first name should be descriptive `fcast_filename`.

DAPPERstats replied:
the issue with shortening to get them all on one line is that the aic file's path name generation is longer (due to the addition of "_model_aic.csv")
as a result `filename_suffix` would have to be shortened to 3 characters, given everything else

given that, I'm going to go with the more descriptive object names

ethanwhite commented on tools/model_tools.R:
```diff
@@ -0,0 +1,39 @@
+#' @param abundance data table with new moon column
+#' @return data table of interpolation-inclusive counts for each species and 
+#'  total
+
```
No blank line here.

ethanwhite commented on tools/model_tools.R:
```diff
@@ -0,0 +1,39 @@
+#' Interpolate missing (rodent) abundance data
+#' 
```
Provide some detail here on how the interpolation is done or change the first line to something like "Replace missing data using linear interpolation"

ethanwhite commented on tools/model_tools.R:
```diff
@@ -0,0 +1,39 @@
+
+      species <- c("BA", "DM", "DO", "DS", "NA.", "OL", "OT", "PB", "PE", 
+                   "PF", "PH", "PL", "PM", "PP", "RF", "RM", "RO", "SF",
+                   "SH", "SO")
```
Typing these in manually is fragile and confusing (I assumed this was an attempt to filter things for quite a while). I'm guessing this is actually just to get the species IDs, but in that case we're missing "PI".

It's better to extract this from the data itself:

```
species <- colnames(all)[2:(ncol(all)-4)]
```

This is still not ideal, which points towards an issue with the main data structure (we should use long, not wide, which will give us a column, named `species` to work with), but that's something to be fixed separately.

DAPPERstats replied:
totally with you on this one.

I grabbed that list from one of the species-level models scripts and just ported it into these ones. That's a bit troubling that we've got a missing species then?

gmyenni replied:
* PI is missing on purpose. It probably doesn't exist at the site, and should not have predictions made.
* Would it not be easier to just switch the data to long format now? Data-wise, it's mostly just a matter of adding `shape=flat` to the abundance function call. Then the models can be written that way from the start. It seems like a waste of time to merge them to work on wide data, then have to change them after the fact? Though, I know we don't want this PR to get too crazy. I could also submit a separate PR, before these go in?

gmyenni replied:
Oh, also, filtering out PI can be moved to the get data step in the process.

DAPPERstats replied:
oh! so do we not want PI include in the totals then, either?
and in that case, should PI not be getting pulled into the rodent data?

DAPPERstats replied:
and for the long vs wide...i'm inclined to hold off of that until we're through with what's being updated in these PRs and then I can double back and update the data format.

DAPPERstats replied:
where would I add the call to not include PI?

gmyenni replied:
In tools/model_functions.R `get_rodent_data`. (This is also where `shape=flat` would get added to `abundance()`.)

Right now, PI still gets included in the totals, because there actually were that many total target rodents in traps. It's just a species ID issue.

DAPPERstats replied:
solid thanks for pointing me there!

DAPPERstats replied:
as for including PI, with how the interpolation goes and due to Jensen's Inequality, the interpolation of the total is not equal to the total of the interpolations. That means, we should either [1] consider all of the species together and just interpolate the total for the four models (i.e. not fit species-specific models) or [2] restrict "total" to mean the total from the list of species included

gmyenni replied:
I think [2] is definitely fine. The interpretation of the `PI` ID (during the period of time it was used at all) should be "unknown Chaetodipus, ~1% chance it's actually a PI, probably a PP." And we already don't include other unknown Chaetodipus in the total, even though they count as target rodents too.

DAPPERstats replied:
ok great, thanks for your insight! 
I'll work on getting this implemented and let you know if I hit any more snags

ethanwhite reviewed (commented):
Looks really good. I've added some more thoughts on style in-line and one here:

Typically we want to import something in the files in which it is used directly. Given this, it seems like `model_tools.R` should be sourced in `esss.R` not `PortalForecasts.R` or `PortalHindcasts.R`. (but I may be missing something here)

juniperlsimonis committed:
move model_tools call into the model script

out of the forecast and hindcast scripts

juniperlsimonis committed:
formatting editing

data frame processing
minor text edits

juniperlsimonis committed:
naming edits

just renaming some things with the output files

juniperlsimonis committed:
edits to the interpolation function

now grabbing species names from the inputted data
this now includes PI, which had been missing before

other minor text edits

juniperlsimonis committed:
updating the data prep to remove PI

as in the remove PI branch

DAPPERstats commented:
i've made all the changes you requested @ethanwhite 
i also updated the data input file to include the removal of PI (i.e., i ported in the updated script and data from #211) to make sure there weren't any problems being introduced by that

everything has passed checks, so it should be good, unless y'all notice anything else

ethanwhite commented:
Looks awesome @DAPPERstats!
